webpack-watcher.js

```diff
@@ -0,0 +1,54 @@
+const childProcess = require('child_process');
+
+module.exports = (options) => {
+   console.log("Forking " + options.username);
+   const child = childProcess.fork(__dirname + '/webpack-compiler.js')
+
+   let onBundled = () => {}
+   const watchers = new Set();
+
+   child.on('message', (message) => {
+      switch (message && message.event) {
+         // {err, stats}
+         case 'built':
+            onBundled(message.err, message.stats);
+            notifyWatchers(message.err);
+            break;
+
+         // If we're not running (building), then there's nothing to wait on.
+         case 'notRunning':
+            notifyWatchers();
+            break;
+      }
+   });
+
+   function notifyWatchers(err) {
+      if (watchers.length) {
+         const success = err ? " has failed" : " has succeeded";
+         console.log("notifying " + watchers.length + " watchers that bundling " + options.username + success);
+      }
+      watchers.forEach(({resolve, reject}) => err ? reject(err) : resolve());
+      watchers.clear();
+   }
+
+   const watchController = {
+      close: () => {
+         console.log("Killing " + options.username);
+         child.kill();
+         notifyWatchers("Webpack killed before building completed")
+      },
+      whenDone: () => new Promise((resolve, reject) => {
+         watchers.add({resolve, reject});
+         child.send({event: 'isRunning'})
+         console.log("asking isRunning");
```

Comment:
Something feels weird to me about the `isRunning` message, but I'm having trouble putting my finger on it. It seems somehow superfluous or mismatched to what we're trying to accomplish.

Could we just maintain our own internal state for "asked for build" and "build done" and refer to that in `whenDone` to decide whether to `notifyWatchers()`?

Reply:
The problem is there is no "build started" event that we can listen for to know the state, all we can do is ask "is running". We could poll it fast and let the parent know when it changes, but that's just more code. I'm open to it, but I'm not sure it's a better solution

Reply:
Ah, I see. We know when we ask it to start watching, and we know when it finishes building something, but it doesn't tell us when it's started building something. That's a bummer.